### PR TITLE
Fast Forwarding Paxos Timestamp Bound Store

### PR DESCRIPTION
**Pull request by jeremyk-91**
Changes:

* Add `ManagedTimestampService` interface extending both TimestampService and TimestampManagementService. We **don't** replace this in PersistentTimestampService and friends because this messes with the pathing in Dropwizard's test infrastructure, it seems.
* Add delegating implementation of the above plus unit tests.
* Change `PaxosTimeLockServer` to create a delegating `ManagedTimestampService`. This is important for ensuring that behind an `InvalidatingLeaderProxy` they both refer to the *same* `PersistentTimestampService`. In the long run #1482 is probably a neater way to fix this, but I think in the short run it's a bit costlier.
* Integration tests for PaxosTBS which look at interactions between the regular timestamp service and managed timestamp service. (Bonus: removed a bit of duplication from the test code.)

This is a key piece of the migration from embedded -> external services workflow.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1484)

### CODE REVIEW

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java [hunk 1]
```diff
@@ -143,6 +148,51 @@ public void timestampServiceShouldRespectDistinctClientsWhenIssuingTimestamps()
...
+        timestampManagementService.fastForwardTimestamp(newTimestamp);
+        assertThat(timestampService.getFreshTimestamp()).isGreaterThan(newTimestamp);
+    }
+
+    @Test
+    public void timestampManagementServiceRespectsTimestampService() {
+        TimestampService timestampService = getTimestampService(CLIENT_1);
+        TimestampManagementService timestampManagementService = getTimestampManagementService(CLIENT_1);
+
+        long newTimestamp = timestampService.getFreshTimestamp() + ONE_MILLION;
```
Comment (hsaraogi) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: nit: `currentTimestampIncrementedByOneMillion`

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java [hunk 2]
```diff
@@ -143,6 +148,51 @@ public void timestampServiceShouldRespectDistinctClientsWhenIssuingTimestamps()
...
+        timestampManagementService.fastForwardTimestamp(newTimestamp);
+        for (int i = 0; i < FORTY_TWO; i++) {
```
Comment (hsaraogi) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: extract to `getFortyTwoFreshTimestamps`

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java [hunk 3]
```diff
@@ -143,6 +148,51 @@ public void timestampServiceShouldRespectDistinctClientsWhenIssuingTimestamps()
...
+            timestampService.getFreshTimestamp();
+        }
+        timestampManagementService.fastForwardTimestamp(newTimestamp + 1);
+        assertThat(timestampService.getFreshTimestamp()).isGreaterThan(newTimestamp + FORTY_TWO);
+    }
+
+    @Test
+    public void fastForwardRespectsDistinctClients() {
+        TimestampService timestampService = getTimestampService(CLIENT_1);
+        TimestampManagementService differentClientTimestampManagementService = getTimestampManagementService(CLIENT_2);
```
Comment (hsaraogi) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: `differentClientTimestampManagementService` -> `anotherClientTimestampManagementService`

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java [hunk 4]
```diff
@@ -143,6 +148,51 @@ public void timestampServiceShouldRespectDistinctClientsWhenIssuingTimestamps()
...
+        assertEquals(currentTimestamp + 1, timestampService.getFreshTimestamp());
+    }
+
+    @Test
+    public void fastForwardToThePastDoesNothing() {
+        TimestampService timestampService = getTimestampService(CLIENT_1);
+        TimestampManagementService timestampManagementService = getTimestampManagementService(CLIENT_1);
+
+        long currentTimestamp = timestampService.getFreshTimestamp();
+        timestampManagementService.fastForwardTimestamp(currentTimestamp + TWO_MILLION);
```
Comment (hsaraogi) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: `currentTimestamp + TWO_MILLION` extract to `currentTimestampIncrementedByTwoMillion`.

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java [hunk 5]
```diff
@@ -143,6 +148,51 @@ public void timestampServiceShouldRespectDistinctClientsWhenIssuingTimestamps()
     }
 
     @Test
+    public void timestampServiceRespectsTimestampManagementService() {
+        TimestampService timestampService = getTimestampService(CLIENT_1);
```
Comment (gsheasby) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: This seems to happen in each test - could it be done in a setup method?

File: atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimeLockServer.java [hunk 6]
```diff
@@ -177,14 +170,19 @@ private TimestampService createPaxosBackedTimestampService(String client) {
...
-                TimestampService.class,
-                () -> PersistentTimestampService.create(
-                        new PaxosTimestampBoundStore(
-                                proposer,
-                                paxosResource.getPaxosLearner(client),
-                                ImmutableList.copyOf(acceptors),
-                                ImmutableList.copyOf(learners),
-                                paxosConfiguration.maximumWaitBeforeProposalMs())),
+                ManagedTimestampService.class,
+                () -> {
```
Comment (gsheasby) [inline atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimeLockServer.java]: This is a big ol' lambda - how about a method for it?

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java (hunk 2 above)
Comment (jeremyk-91) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: Sure, done.

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java (hunk 1 above)
Comment (jeremyk-91) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: Changed (and changed in the test going the other way too.)

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java (hunk 3 above)
Comment (jeremyk-91) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: OK, changed.

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java (hunk 4 above)
Comment (jeremyk-91) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: OK, changed.

File: atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimeLockServer.java (hunk 6 above)
Comment (jeremyk-91) [inline atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimeLockServer.java]: Sure, created.

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java (hunk 5 above)
Comment (jeremyk-91) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java]: Actually the timestamp service (and management service) for client 1 can be private final fields of the class, so I've made them that.

**Comment by jeremyk-91** [comment]
@hsaraogi Yes. If this was not the case, then the `PaxosTimeLockServerIntegrationTest` will fail (we have tests that query both the timestamp and timestamp management services, and these work via proxies that query the relevant endpoints).

**Pull request merged**